libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+        return retVal.toTypedArray()
+    }
+
+    private fun parseRouteProgressAsync(): Deferred<TelemetryStep> {
+        return jobControl.scope.async {
+            val telemetryStep = TelemetryStep() // Default initialization
+            val routeProgress = channelOnRouteProgress.receive()
+            routeProgress.currentLegProgress()?.upcomingStep()?.let { legStep ->
+                telemetryStep.distance = legStep.distance().toInt()
+                telemetryStep.distanceRemaining = routeProgress.distanceRemaining().toInt()
+                telemetryStep.duration = legStep.duration().toInt()
+                telemetryStep.durationRemaining = routeProgress.durationRemaining().toInt()
+                telemetryStep.previousModifier = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.modifier()
+                        ?: ""
+                telemetryStep.previousName = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.name()
+                        ?: ""
+                telemetryStep.previousType = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.type()
+                        ?: ""
+                telemetryStep.upcomingType = legStep.maneuver().type() ?: ""
+                telemetryStep.upcomingModifier = legStep.maneuver().modifier() ?: ""
+                telemetryStep.upcomingName = legStep.name() ?: ""
+            }
+            telemetryStep
+        }
+    }
+
+    /**
+     * TODO:OZ add one of these for each telemetry event
+     * Generates a well-formed Feedback event.
+     */
+    private suspend fun populateUserFeedbackEvent(event: TelemetryEventFeedback): TelemetryEventInterface {
+        val phoneState = PhoneState(context)
+
+        val feedbackEvent = TelemetryUserFeedbackWrapper(event,
+                phoneState.userId,
+                phoneState.audioType,
+                getDatePartitionedLocations { location -> location.time < System.currentTimeMillis() },
+                getDatePartitionedLocations { location -> location.time > System.currentTimeMillis() },
+                UUID.randomUUID().toString(),
+                event.screenShot,
+                parseRouteProgressAsync().await()
+        )
+        return feedbackEvent
+    }
+
+    /**
+     * TODO:OZ add code to handle all other event types. Once implmented, instead of throwing an exception the code
+     * should log an error
+     */
+    private suspend fun populateTelemetryEventWrapper(event: TelemetryEventInterface): TelemetryEventInterface? =
+            when (event) {
+                is TelemetryEventFeedback -> populateUserFeedbackEvent(event)
+                else -> throw(NavigationException("Unsupported telemetry event"))
+            }
+
+    private fun validateAccessToken(token: String?) {
+        token?.let { accessToken ->
+            if (accessToken.isEmpty() || !accessToken.toLowerCase(Locale.US).startsWith("pk.") && !accessToken.toLowerCase(
```

Comment:
`let` could be removed, take a look at second logic operator `&&` ->`||`
```suggestion
            if (accessToken.isNullOrEmpty() || !accessToken.toLowerCase(Locale.US).startsWith("pk.") || !accessToken.toLowerCase( Locale.US).startsWith("sk."){
            throw NavigationException("A valid access token must be passed in when first initializing MapboxNavigation")
            }
```

Reply:
👍

Reply:
👍